hourly-tips/src/main/java/org/apache/flink/training/exercises/hourlytips/HourlyTipsExercise.java

```diff
@@ -72,22 +79,38 @@ public JobExecutionResult execute() throws Exception {
         // set up streaming execution environment
         StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
 
+        WatermarkStrategy<TaxiFare> strategy = WatermarkStrategy
+                .<TaxiFare>forMonotonousTimestamps()
+                .withTimestampAssigner((event, timestamp) -> event.getEventTimeMillis());
+
         // start the data generator
-        DataStream<TaxiFare> fares = env.addSource(source);
+        DataStream<TaxiFare> fares = env.addSource(source)
+                .assignTimestampsAndWatermarks(strategy);
 
-        // replace this with your solution
-        if (true) {
-            throw new MissingSolutionException();
-        }
+        fares.keyBy(fare -> fare.driverId)
+                .window(TumblingEventTimeWindows.of(Time.hours(1)))
+                .process(new AddTips())
```

Comment:
Looks OK
How would a solution using `window().reduce()` work?

Reply:
I looked into it a bit, I think I would map the fare event into a tuple and create a reduce method to add the tips for each driver. I can try and make an implementation.

Reply:
Implemented a solution using reduce.